runtime/caml/mlvalues.h

```diff
@@ -135,8 +135,12 @@ bits  63        (64-P) (63-P)        10 9     8 7   0
 #define Profinfo_hd(hd) NO_PROFINFO
 #endif /* WITH_PROFINFO */
 
-#define Hd_val(val) (((header_t *) (val)) [-1] + 0)
 #define Hp_atomic_val(val) ((atomic_uintnat *)(val) - 1)
+#define Hd_val(val) ((header_t) \
+  (atomic_load_explicit( \
+    (Hp_atomic_val(val)), \
+    memory_order_relaxed)))
```

Comment:
It might fit on two lines only?
```suggestion
#define Hd_val(val) ((header_t) \
  (atomic_load_explicit(Hp_atomic_val(val), memory_order_relaxed)))
```